File: Kotlin/src/TAD.kt
package bst.tad

class Node(key: Int){

    var key : Int = key
    var left : Node? = null
    var right : Node? = null
}

//------------------------Node Creation--------------------------//
fun createNode(key: Int): Node{
    return Node(key)
}

//------------------------Empty Tree-----------------------------//
fun emptyTree(): Node?{
    return null
} 

//------------------------ Recursive Insert----------------------//
fun insert_r(abb_in: Node?, key: Int): Node?{
    var abb = abb_in

	if (abb == null){
		abb = createNode(key)
    }else{
		if (key < abb.key)
			abb.left = insert_r(abb.left,key)
        else if (key > abb.key)
			abb.right = insert_r(abb.right,key)
    }
    return abb
}

//------------------------ Iterative Insert----------------------//
fun insert_i(abb_in: Node?,key: Int): Node?{
	var new_node = createNode(key)
	var abb = abb_in
    var son : Node?
    var father : Node?

	if (abb == null){
		return new_node			
    }else{
		son = abb
		father = null
		while ((son != null) and (son?.key != key)){
			father = son
			if (key < son!!.key){
				son = son.left
            }else{
				son = son.right
            }
        }		
		if (son == null){
			if (key < father!!.key){
				father.left = new_node
            }else{
				father.right = new_node
            }
        }
		return abb
    }
}

//-----------------------------Insert key------------------------//
fun insertKey(abb: Node?,key: Int): Node?{
    return insert_i(abb,key)
}

//------------------------ Recursive Search----------------------//
fun search_r(abb_in: Node?,key: Int): Node?{
    var abb = abb_in

	if (abb == null)
		return null
	else if (abb.key == key)
		return abb
	else if (key < abb.key)
		return search_r(abb.left,key)
	else
		return search_r(abb.right,key)
}

//------------------------ Iterative Search----------------------//
fun search_i(abb_in: Node?,key: Int): Node?{
	var node = abb_in

	while ((node != null) and (node!!.key != key)){
		if (key < node.key)
			node = node.left
		else
			node = node.right
    }
	return node
}

//----------------------------Search key-------------------------//
fun	searchKey(abb: Node?,key: Int): Node?{
	return search_r(abb,key)
}	

//-----------------------------Left Son--------------------------//	
fun leftSon(abb: Node?): Node?{
	return abb?.left
}
		
//----------------------------Right Son--------------------------//
fun rightSon(abb: Node?): Node?{
	return abb?.right
}
		
//-------------------------------Key-----------------------------//	
fun key(abb: Node?): Int{
	return abb!!.key
}
		
//----------------------------Left Son---------------------------//	
fun isEmptyTree(abb: Node?): Boolean{
	return (abb == null)
}

//------------------------ Recursive Erase-----------------------//
fun erase_r(abb_in: Node?,key: Int): Node?{
    var abb = abb_in
	var aux: Node?
    aux = abb

	fun sup2(tree_in: Node?): Node?{
        var tree = tree_in

		if (tree?.right != null)
			tree.right = sup2(tree.right)
		else{
			aux?.key = tree!!.key
			return tree.left
        }
		return aux
    }

	if (abb != null){
		if (key < abb.key)
			abb.left = erase_r(abb.left,key)
		else if (key > abb.key)
			abb.right = erase_r(abb.right,key)
		else{
			aux = abb
			if (abb.left == null)
				abb = abb.right
			else if (abb.right == null)
				abb = abb.left
			else
				abb = sup2(abb.left)
        }
    }
	return abb
}

//------------------------- Iterative Erase----------------------//
fun erase_i(abb_in: Node?,key: Int): Node?{
    var abb = abb_in
	var sons_num: Int
	var sup: Node?
	var f_sup: Node? = null
	var not_empty_son: Node?
	var son_left_max: Node?
		
	sup = abb
	while ((sup != null) and (sup?.key != key)){
		f_sup = sup
		if (key < sup!!.key)
			sup = sup.left
		else
			sup = sup.right
    }		
	if (sup != null){
		sons_num = 0
		if (sup.left != null)
			sons_num = sons_num + 1
		if (sup.right != null)
			sons_num = sons_num + 1
				
		if (sons_num == 0){
			if (f_sup == null)
				abb = null
			else if (f_sup.left == sup)
				f_sup.left = null
			else
				f_sup.right = null
					
        }else if (sons_num == 1){
			if (sup.left == null)
				not_empty_son = sup.right
			else	
				not_empty_son = sup.left
				
			if (f_sup == null)
				abb = not_empty_son
			else if (f_sup.left == sup)
				f_sup.left = not_empty_son
			else
				f_sup.right = not_empty_son
				
        }else{
			f_sup = sup
			son_left_max = sup.left
			while (son_left_max?.right != null){
				f_sup = son_left_max
				son_left_max = son_left_max.right
            }	
			sup.key = son_left_max!!.key
			if (f_sup == sup)
				f_sup.left = son_left_max.left
			else
				f_sup?.right = son_left_max.left
				
        }
    }	
	return abb
}

//---------------------------- Erase Key-------------------------//				
fun eraseKey(abb: Node?,key: Int): Node?{
	return erase_i(abb,key)
}

File: Python/TAD.py
#
#***** Autor1: Yeray Méndez Romero
#***** Login1: yeray.mendez
#***** Email1: yeray.mendez@udc.es
#***** Autor2: Daniel Rivera López
#***** Login2: d.rivera1
#***** Email2: d.rivera1@udc.es
# 

class Node:
	def __init__(self, key):
		self.key = key
		self.left = None
		self.right = None

def createNode(key):
#CreateNodeA(key)->Node(Key,None,None)
#***** Goal: Create a new node.
#***** Input: key.
#***** OutPut: Node(Key,None,None).
#***** PreCondition:
	return Node(key)

def emptyTree(abb):
#Empty Tree(abb)-> None
#***** Goal: Create a empty tree.
#***** Input: Binary Tree
#***** OutPut: None
#***** PreCondition:
	return None

def insert_r(abb,key):
#insert_r(abb,key)-> Tree with new Key.
#***** Goal: Insert key in the right place in the tree.
#***** Input: Binary tree and key.
#***** OutPut: Key inserted in the tree.
#***** PreCondition: Initialized tree.
	if abb is None:
		return createNode(key)
	else:
		if key < abb.key:
			abb.left = insert_r(abb.left,key)
		elif key > abb.key:
			abb.right = insert_r(abb.right,key)
	return abb

def insert_i(abb,key):
#insert_i(abb,key)-> Tree with new Key.
#***** Goal: Insert key in the right place in the tree.
#***** Input: Binary tree and key.
#***** OutPut: Key inserted in the tree.
#***** PreCondition: Initialized tree.
	new_node = createNode(key)
	
	if abb is None:
		return new_node			
	else :
		son = abb
		father = None
		while son is not None and son.key != key:
			father = son
			if key < son.key :
				son = son.left
			else :
				son = son.right
				
		if son is None :
			if key < father.key :
				father.left = new_node
			else :
				father.right = new_node
		return abb

def	insertKey(abb,key):
#insertKey(abb,key)-> Tree with new Key.
#***** Goal: Insert key in the right place in the tree.
#***** Input: Binary tree and key.
#***** OutPut: Key inserted in the tree.
#***** PreCondition: Initialized tree.
	return insert_r(abb,key)
		
def search_r(abb,key):
#search_r(abb,key)-> Node that contain the key.
#***** Goal: Return the node which contains the key.
#***** Input: Binary tree and key.
#***** OutPut: Node that contains the key.
#***** PreCondition: Initialized tree.
	if abb is None:
		return None
	elif abb.key == key:
		return abb
	elif key < abb.key:
		return search_r(abb.left,key)
	else:
		return search_r(abb.right,key)

def search_i(abb,key):
#search_i(abb,key)-> Node that contain the key.
#***** Goal: Return the node which contains the key.
#***** Input: Binary tree and key.
#***** OutPut: Node that contains the key.
#***** PreCondition: Initialized tree.
	node = abb
	while node is not None and node.key != key:
		if key < node.key:
			node = node.left
		else:
			node = node.right
	return node

#------------------------Search key----------------------#	
def	searchKey(abb,key):
#searchKey(abb,key)-> Node that contain the key.
#***** Goal: Return the node which contains the key.
#***** Input: Binary tree and key.
#***** OutPut: Node that contains the key.
#***** PreCondition: Initialized tree.
	return search_i(abb,key)	

def leftSon(abb):
#leftSon(abb)-> Left son of the tree.
#***** Goal: Returns the left branch of the tree.
#***** Input: Binary tree not null.
#***** OutPut: Returns the left branch of the tree.
#***** PreCondition: Initialized tree.
	return abb.left

def rightSon(abb):
#rightSon(abb)-> Right son of the tree.
#***** Goal: Returns the right branch of the tree.
#***** Input: Binary tree not null.
#***** OutPut: Returns the left branch of the tree.
#***** PreCondition: Initialized tree.
	return abb.right

def key(abb):
#key(abb)-> Root.
#***** Goal: Returns the root of the tree.
#***** Input: Binary tree not null.
#***** OutPut: Returns the root.
#***** PreCondition:Initialized tree.
	return abb.key

def isEmptyTree(abb):
#isEmptyTree(abb)-> Boolean.
#***** Goal: Indicates if the tree is empty.
#***** Input: Binary tree.
#***** OutPut: Boolean.
#***** PreCondition:Initialized tree.
	return abb is None

def erase_r(abb,key):
#erase_r(abb,key)-> Tree without the key.
#***** Goal: Remove the key from the tree.
#***** Input: Binary tree and key.
#***** OutPut: Tree without the key.
#***** PreCondition:Initialized tree.

	aux = None

	def sup2(tree):
		if tree.right is not None:
			tree.right = sup2(tree.right)
		else:
			aux.key = tree.key
			#aux = tree
			return tree.left
		return aux

	if abb is not None:
		if key < abb.key:
			abb.left = erase_r(abb.left,key)
		elif key > abb.key:
			abb.right = erase_r(abb.right,key)
		else:
			aux = abb
			if abb.left is None:
				abb = abb.right
			elif abb.right is None:
				abb = abb.left
			else:
				abb = sup2(abb.left)
			aux = None
	return abb
			
def erase_i(abb,key):
#erase_i(abb,key)-> Tree without the key.
#***** Goal: Remove the key from the tree.
#***** Input: Binary tree and key.
#***** OutPut: Tree without the key.
#***** PreCondition:Initialized tree.
	sons_num = None
	sup = None
	f_sup = None
	not_empty_son = None
	son_left_max = None
		
	f_sup = None
	sup = abb
	while sup is not None and sup.key != key:
		f_sup = sup
		if key < sup.key:
			sup = sup.left
		else:
			sup = sup.right
				
	if sup is not None:
		sons_num = 0
		if sup.left != None:
			sons_num = sons_num + 1
		if sup.right != None:
			sons_num = sons_num + 1
				
		if sons_num == 0:
			if f_sup == None:
				abb = None
			elif f_sup.left == sup:
				f_sup.left = None
			else:
				f_sup.right = None
					
		elif sons_num == 1:
			if sup.left == None:
				not_empty_son = sup.right
			else:	
				not_empty_son = sup.left
				
			if f_sup == None:
				abb = not_empty_son
			elif f_sup.left == sup:
				f_sup.left = not_empty_son
			else:
				f_sup.right = not_empty_son
				
		else:
			f_sup = sup
			son_left_max = sup.left
			while son_left_max.right != None:
				f_sup = son_left_max
				son_left_max = son_left_max.right
				
			sup.key = son_left_max.key
			if f_sup == sup:
				f_sup.left = son_left_max.left
			else:
				f_sup.right = son_left_max.left
				
			sup = son_left_max

		sup = None	
	return abb

def eraseKey(abb,key):
#eraseKey(abb,key)-> Tree without the key.
#***** Goal: Remove the key from the tree.
#***** Input: Binary tree and key.
#***** OutPut: Tree without the key.
#***** PreCondition:Initialized tree.
	return erase_i(abb,key)